include/circt/Support/PrettyPrinter.h

```diff
@@ -0,0 +1,298 @@
+//===- PrettyPrinter.h - Pretty printing ----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This implements a pretty-printer.
+// "PrettyPrinting", Derek C. Oppen, 1980.
+// https://dx.doi.org/10.1145/357114.357115
+//
+// This was selected as it is linear in number of tokens O(n) and requires
+// memory O(linewidth).
+//
+// See PrettyPrinter.cpp for more information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_PRETTYPRINTER_H
+#define CIRCT_SUPPORT_PRETTYPRINTER_H
+
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/SaveAndRestore.h"
+
+#include <cstdint>
+#include <deque>
+#include <limits>
+
+namespace circt {
+namespace pretty {
+
+//===----------------------------------------------------------------------===//
+// Tokens
+//===----------------------------------------------------------------------===//
+
+/// Style of breaking within a group:
+/// - Consistent: all fits or all breaks.
+/// - Inconsistent: best fit, break where needed.
+/// - Never: force no breaking including nested groups.
+enum class Breaks { Consistent, Inconsistent, Never };
+
+/// Style of indent when starting a group:
+/// - Visual: offset is relative to current column.
+/// - Block: offset is relative to current base indentation.
+enum class IndentStyle { Visual, Block };
+
+class Token {
+public:
+  enum class Kind { String, Break, Begin, End };
+  struct StringInfo {
+    const char *str;
+    uint32_t len;
+  };
+  struct BreakInfo {
+    uint32_t spaces; // How many spaces to emit when not broken.
+    int32_t offset;  // How many spaces to emit when broken.
+    bool neverbreak; // If set, behaves like break except this always 'fits'.
+  };
+  struct BeginInfo {
+    int32_t offset; // Adjust base indentation by this amount.
+    Breaks breaks;
+    IndentStyle style;
+  };
+  struct EndInfo {
+    // Nothing
+  };
+
+private:
+  union {
+    StringInfo stringInfo;
+    BreakInfo breakInfo;
+    BeginInfo beginInfo;
+    EndInfo endInfo;
+  } data;
+  Kind kind;
+
+protected:
+  template <Kind k, typename T>
+  static auto &getInfoImpl(T &t) {
+    if constexpr (k == Kind::String)
+      return t.data.stringInfo;
+    if constexpr (k == Kind::Break)
+      return t.data.breakInfo;
+    if constexpr (k == Kind::Begin)
+      return t.data.beginInfo;
+    if constexpr (k == Kind::End)
+      return t.data.endInfo;
+    llvm_unreachable("unhandled token kind");
+  }
+
+  Token(Kind k) : kind(k) {}
+
+public:
+  Kind getKind() const { return kind; }
+};
+
+/// Helper class to CRTP-derive common functions.
```

Comment:
Anytime there is a `classof` (manual implementation of LLVM RTTI, yes?) I wonder if this can be moved into ODS. Can it? Have you considered that?

There is some possible similarity to the factoring of the GCT View Augmented Type stuff into attributes here: https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/FIRRTLAttributes.td#L93

Also, only for your consideration.